client/pages/login.vue

```diff
@@ -202,6 +230,10 @@ export default {
     }
   },
   async mounted() {
+
+    // Check for proxy auth
+    this.checkForwardAuth();
+
     if (localStorage.getItem('token')) {
       var userfound = await this.checkAuth()
```

Comment:
This `checkForwardAuth` function is asynchronous and since it isn't being waited for there may be a token that gets validated and/or the `/status` call being made simultaneously.

Even if we added an `await` this would not be good for servers that haven't been initialized yet. The `/status` request is how we find out if the server has been setup.

I'm not sure how this should be structured but we want to make the `/status` call first. We could move the `checkAuth` to happen after the `/status` that way we can first find out if forward auth is configured. I assume if forward auth is configured we want to use that instead of using the persisted user token.

Reply:
> I'm not sure how this should be structured but we want to make the /status call first. We could move the checkAuth to happen after the /status that way we can first find out if forward auth is configured. I assume if forward auth is configured we want to use that instead of using the persisted user token.

Would it then make sense to implement the Info directly into `/status` ? Given that also other options like the ConfigPath are passed.

> I assume if forward auth is configured we want to use that instead of using the persisted user token.

If I understand the code correctly, I think it can be done in https://github.com/advplyr/audiobookshelf/blob/0e292c64c44c3dd0af4cd5f0338bdda23b90c620/server/Auth.js#L55

One could copy the `this.user.find` and feed it directly with the value from the username header. However the problem is that the JWT verifier here also uses the userID to find an user: https://github.com/advplyr/audiobookshelf/blob/0e292c64c44c3dd0af4cd5f0338bdda23b90c620/server/Auth.js#L112
Not sure if this required from ABS. Authentik can also supply a UserID (just a normal unsigned integer) over the headers, but most software implementing Forward Auth only rely on either unique usernames or unique e-mail-adresses.

Reply:
Actually if we still use the auth token in Abs then we don't need to update the `authMiddleware`.
Here is what I'm thinking for the login page.

If a token is stored then `checkAuth` and login using the token.
Otherwise call `/status` which will:
 - If the server is not initialized it responds how it does now
 - If the PROXY_FORWARD env variables are set then attempt to get/create the user. In this case the user will be returned in with the response and the web ui can just `setUser`
 - Else regular login is requested

With this method if you are already using token login before setting this up you will need to logout first.

Your comment on the logout button also made me realize we would need to store a flag on the User if it is created or authenticated at all via the FORWARD_PROXY. This flag is how we could tell whether the logout button needs an additional step. How would logout be implemented for Authentik?
Users created from the forward proxy won't have passwords set so we wouldn't want them using the regular login form.

Reply:
This would be weird if the server hasn't been initialized yet.

Reply:
> Here is what I'm thinking for the login page.
> 

Sounds good, I will wait a bit until the PR creator (or you) started work on it as I cannot do commits on this PR and don't want to directly hijack with a new PR.  


> Your comment on the logout button also made me realize we would need to store a flag on the User if it is created or authenticated at all via the FORWARD_PROXY. This flag is how we could tell whether the logout button needs an additional step. How would logout be implemented for Authentik? Users created from the forward proxy won't have passwords set so we wouldn't want them using the regular login form.

Actually this also should be configurable with a ENV variable, as every SSO software has an potentially different URL for that. On Authentik the default value for that is `/outpost.goauthentik.io/sign_out` (which is a special location directive in nginx proxied to authentik instead to ABS). This could be also delivered via `/status` so the Button can be adjusted. 
When the user clicks on it, he gets forwarded to `audiobookshelf.example.com/outpost.goauthentik.io/sign_out`, authentik clears his session, and wont allow the user on audiobookshelf any more (and gives him then the option to log off from authentik completely or to log back in into ABS).

> store a flag on the User

Temporarily for handling it in the UI? Yes that sounds smart. However I wouldn't store that permanently in the DB, so that one can again deactivate Forward Auth.

> Users created from the forward proxy won't have passwords set so we wouldn't want them using the regular login form

A good way is also to simply set a random password for the user on creation or to not allow a login with an empty password. This way the feature can be completely disabled again while still keeping the users in tact and without creating a security issue.

And the password change function should be probably disabled, too. This could also be then done by just checking wether the user is currently logged in via SSO.

---

If you want to make it even more advanced. One can also implement the header for the SSO Groups (Authentik provides a list via the X-authentik-groups header). 
Example:
```
export PROXY_FORWARD_AUTH_GROUP_HEADER=X-authentik-groups
export PROXY_FORWARD_AUTH_GROUP_ADMIN=AudioBookAdmin
export PROXY_FORWARD_AUTH_GROUP_USER=AudioBookUser
export PROXY_FORWARD_AUTH_GROUP_GUEST=AudioBookGuest
```

This way a user is set automatically to admin if he has in authentik the group AudioBookAdmin. But this is not soo important and probably shouldnt be done in the first PR.

Reply:
> > store a flag on the User
> 
> Temporarily for handling it in the UI? Yes that sounds smart. However I wouldn't store that permanently in the DB, so that one can again deactivate Forward Auth.

I think I'm following everything so far except how we would detect whether a user is signed in via SSO when they are logging in via a persisted token.

Back to my post earlier I thought we would just use the persisted token if it existed.

> If a token is stored then `checkAuth` and login using the token. Otherwise call `/status` which will:

In this case I guess we could add an additional step in the `/api/authorize` endpoint that checks if the auth forward is configured then if the username from the token matches the user retrieved from the token.

Reply:
> In this case I guess we could add an additional step in the /api/authorize endpoint that checks if the auth forward is configured then if the username from the token matches the user retrieved from the token.

Ah yes, yep that seems to be indeed necessary. And should work this way; otherwise one ignores the header. The endpoint can then like you suggest also return if the user is logged in via SSO (when configured and header and token matches).